src/main/java/com/squareup/javapoet/AnnotationSpec.java

```diff
@@ -183,13 +185,17 @@ public Builder toBuilder() {
   }
 
   @Override public boolean equals(Object o) {
-    return o instanceof AnnotationSpec
-        && ((AnnotationSpec) o).type.equals(type)
-        && ((AnnotationSpec) o).members.equals(members);
+    if (this == o)
+      return true;
+    if (o == null)
+      return false;
+    if (getClass() != o.getClass())
+      return false;
+    return hashCode == o.hashCode();
```

Comment:
That's what it does - with only a 32 bit foot print and not computing `toString()` (with `emit()` etc) everytime `equals(Object)` is called. Somehow _disgusting_, sure.
Concerning correctness: i think, according to `String.hashCode()`s Javadoc, it should be nearly 100% correct.

But, if having the `toString()` result cached in each instance is a) not a memory concern and b) the runtime impact of `this.toString().equals(other.toString())` is not a performance concern, I'll rewrite the code accordingly.